### PR TITLE
C API

### PR DESCRIPTION
**Pull request by franziskuskiefer**
This is a C API for the verifier. What do you think?
I have [a separate repo](/franziskuskiefer/cose-verifier-cpp) that uses this API.

When this one is done I think we can clean up a little and start looking at the gecko side.

### CODE REVIEW

File: src/capi.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                                    *const usize, /* cert lengths in cert_chain */
+                                    *const u8, /* signer cert */
+                                    usize, /* signer cert len */
+                                    *const u8, /* signature bytes */
+                                    usize, /* signature len */
+                                    u8 /* signature algorithm */)
+                                    -> bool;
+
+#[no_mangle]
+pub unsafe extern "C" fn verify_signature_with_cpp(
```
Comment (mozkeeler) [inline src/capi.rs]: Does this whole function need to be unsafe?
Also, technically this is a c api, so maybe verify_signature_from_c? or verify_signature_ffi?

File: src/capi.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    if cose_signatures.len() < 1 {
+        return false;
+    }
+
+    let mut result = true;
+    for cose_signature in cose_signatures {
+        let signature_type = cose_signature.signature_type;
+        // ES256 = 0, ES384 = 1, ES521 = 2, PS256 = 3
+        let signature_type = match signature_type {
+            SignatureAlgorithm::ES256 => 0,
```
Comment (mozkeeler) [inline src/capi.rs]: Should we expose some constants in this API so we don't have to use the actual numerical values?

File: src/capi.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            cert_lens.push(cert.len());
+            certs.push(cert.as_ptr());
+        }
+
+        // Call callback to verify the parsed signatures.
+        result &= verify_callback(
+            real_payload.as_ptr(),
+            real_payload.len(),
+            certs.as_ptr(),
+            cose_signature.certs.len(),
```
Comment (mozkeeler) [inline src/capi.rs]: Maybe certs.len() here?

File: src/capi.rs (hunk 2 above)
Comment (franziskuskiefer) [reply src/capi.rs]: I'd love to but I don't think this is possible. We could add a getter function or something. But that's probably more confusing than helping.

File: src/capi.rs (hunk 1 above)
Comment (franziskuskiefer) [reply src/capi.rs]: `_ffi` sounds good

**Comment by franziskuskiefer** [comment]
Closing and merging PR #42.

**Pull request closed without merging**